### PR TITLE
Add ClrGenericParameter

### PR DESCRIPTION
**Pull request by NextTurn**
Contributes to #408

### CODE REVIEW

**Comment by NextTurn** [comment]
`QueryInterface` returns `E_NOINTERFACE` for `IID_IMetaDataImport2` on .NET Framework. Using `IID_IMetaDataImport` as a workaround, since `IMetaDataImport2` is available since .NET Framework 2.0.

File: src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdType.cs [hunk 1]
```diff
@@ -124,19 +124,30 @@ private GCDesc GetOrCreateGCDesc()
...
 
+        public override IEnumerable<ClrGenericParameter> EnumerateGenericParameters()
+        {
+            MetaDataImport2? import = Module?.MetadataImport2;
+            if (import is null)
+                yield break;
+
+            foreach (int token in import.EnumerateGenericParams((int)MetadataToken))
+                if (import.GetGenericParamProps(token, out int index, out GenericParameterAttributes attributes, out string? name))
+                    yield return new ClrGenericParameter(index, attributes, name);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdType.cs]: We should add the token itself to ClrGenericParameter.

File: src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj [hunk 2]
```diff
@@ -4,9 +4,6 @@
     <TargetFrameworks>netcoreapp3.0</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
 
-    <!-- This _IS_ a Unit Test Project, but because of build_test_assets.cmd, we don't want to run as part of CI right now -->
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj]: This should be in its own pull request.  Currently the tests fail when this block is removed, even on Linux.  I haven't had a chance to dig through it yet.  I agree we should remove this block, but it should be in its own PR with associated fixes.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrGenericParameter.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public readonly struct ClrGenericParameter
+    {
+        public int Index { get; }
+
+        public GenericParameterAttributes Attributes { get; }
+
+        public string Name { get; }
+
+        internal ClrGenericParameter(int index, GenericParameterAttributes attributes, string name)
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrGenericParameter.cs]: We should add the metadata token for this generic parameter in case the user wants to grab more data out of IMetaDataReader in the future.

File: src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            name = null;
+            InitDelegate(ref _getGenericParamProps, VTable.GetGenericParamProps);
+
+            int hr = _getGenericParamProps(
+                Self, token, out index, out attributes, out int owner, out _, null, 0, out int needed);
+
+            if (hr < 0)
+                return false;
+
+            string nameResult = new string('\0', needed - 1);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs]: I like this pattern.  I'll try to use this in the future for other SOS/Dac calls if I get around to it.

File: src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            try
+            {
+                IntPtr handle = IntPtr.Zero;
+                int hr;
+
+                while ((hr = _enumGenericParams(Self, ref handle, token, tokens, tokens.Length, out int count)) >= 0 && count > 0)
+                    for (int i = 0; i < count; i++)
+                        yield return tokens[i];
+
+                CloseEnum(handle);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs]: I'd suggest moving this to the finally block as `if (handle != IntPtr.Zero) CloseEnum(handle);`.

File: src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj (hunk 2 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime.Tests/Microsoft.Diagnostics.Runtime.Tests.csproj]: This isn't part of the PR, but the PR will be blocked until the tests pass.

**Comment by NextTurn** [comment]
I need to call `QueryInterface` on `IMetaDataImport` with `IID_IMetaDataImport2` to get `IMetaDataImport2`, which isn't possible with the current `CallableCOMWrapper` implementation. Shall I just use `IID_IMetaDataImport` based on the fact that the two interfaces are the same (since .NET Framework 2.0)?

**Comment by leculver** [comment]
Sure.  Or change CallableCOMWrapper to expose what it should, whichever you think is better or more efficient for you.  CallableCOMWrapper is intentionally marked public and should support what you are trying to do, but don't feel you need to fix it.

File: src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs (hunk 5 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs]: Then we have duplicated zero checks in both the caller and the callee.

File: src/Microsoft.Diagnostics.Runtime/src/Builders/RuntimeBuilder.cs [hunk 6]
```diff
@@ -1057,7 +1057,7 @@ private void CreateFieldsForMethodTableWorker(ClrType type, out ImmutableArray<C
             statics = staticsBuilder.MoveOrCopyToImmutable();
         }
 
-        public MetaDataImport? GetMetaDataImport(ClrModule module) => _sos.GetMetadataImport(module.Address);
+        public MetaDataImport? GetMetadataImport(ClrModule module) => _sos.GetMetadataImport(module.Address);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Builders/RuntimeBuilder.cs]: I don't mind following the guidelines you point to as long as it's 100% consistent.  I don't want the class name and the method name to not match.  Note that the reason it's MetaDataImport in ClrMD is because that's how [IMetaDataImport](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/metadata/imetadataimport-interface) is spelled.

If if you change `MetaData` to `Metadata` I'm fine with that (I like it better even) but also please change the `MetaDataImport` class with this PR too.

File: src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs (hunk 5 above)
Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/DacInterface/MetaDataImport2.cs]: That's ok, having the CloseEnum in the finally block is worth an extra branch.

File: src/Microsoft.Diagnostics.Runtime/src/Builders/RuntimeBuilder.cs (hunk 6 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/Builders/RuntimeBuilder.cs]: > Note that the reason it's MetaDataImport in ClrMD is because that's how IMetaDataImport is spelled.

I think the consistency with other types and properties (for example, `MetadataToken`) is more important than to follow the interface's old-style real name. I've renamed that type to adopt the new style.
